docs/content/sips/draft-000-templates.md

```diff
@@ -0,0 +1,365 @@
+  write different docs for every language.
+* Having our own system doesn't preclude offering Spin templates via
+  language-idiomatic systems too.
+
+Reasons why we shouldn't:
+
+* It's a decent chunk of work to do it well.
+* It's extra code for every copy of Spin to carry around.
+
+I believe the general consensus is that the case for having templates built
+in is compelling, but it's always worth checking that assumption.
+
+## Identifying a template
+
+Templates are currently identified by a repository ID and directory name.
+Because Spin is a multi-language runtime, it would be good to also identify
+templates by language, so that a user could easily see and choose templates
+in their preferred language.
+
+The question of the repository ID is what we want "adding" a template to mean.
+Does adding a template:
+
+* Record a mapping of a repository ID to a Git location
+* Copy a template to the local system
+
+At the moment it seems to do both, which I found a bit confusing. If we take
+the view that a template is copied, then we no longer need the repository
+ID and can address templates by their names along.
+
+This, by the way, is the approach taken by the `dotnet` templating system,
+where you run `dotnet new --install <package-name>` to install a template
+package, and then use template by running `dotnet new <template-name>`,
+with an optional `-lang` option to specify the language.
+
+Because this does not qualify every template with a unique Git URL, it does
+require some consideration of name clashes.  There may also be value in
+allowing users to run templates without installing them.
+
+## User experience: starting a template
+
+The `spin new` command create a new Spin application.  (We may want to consider
+also allowing it to create a new component in an existing application.)
+
+Proposed syntax:
+
+| Command        | Behaviour |
+|----------------|-----------|
+| `spin new`     | Prompt from installed templates, then run chosen template |
+| `spin new http` | Runs the `http` template. If this is available in more than one language, prompts for language or uses default language |
+| `spin new http -lang haskell` | Runs the Haskell `http` template |
+| `spin new -lang haskell` | Prompt from installed Haskell-language templates, then run chosen template (maybe not useful?) |
+
+### What if there are no templates?
+
+We don't have an installer so we can't guarantee to have templates present the
+first time the user runs `spin new`.  The `cargo generate` experience for this
+is to print an error about a config file path; this is not good.
+
+If the user runs `spin new` (with any options) and we do not have any templates
+installed, we should offer to install the default ones from Spin itself.  We should
+also point them at the `spin templates install` command to add more.
+
+### Template search strategy
+
+If the user runs `spin http foo`:
+
+* If there is NO default language set:
```

Comment:
Is a default language required? Seems like you could deploy e.g. bartholomew without caring what language it is written in.

Reply:
Oh look theres a whole section for it...